File: tools/python/maton_agent_toolkit/shared/mcp_client.py
"""Client for connecting to the Maton MCP server at mcp.maton.ai."""

import json
import os
import warnings
from contextlib import asynccontextmanager
from typing import List, Dict, Any, AsyncGenerator
from typing_extensions import TypedDict

from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client

from .async_initializer import AsyncInitializer
from .constants import VERSION, MCP_SERVER_URL, TOOLKIT_HEADER, MCP_HEADER


class McpToolInputSchema(TypedDict, total=False):
    """JSON Schema for MCP tool input."""

    type: str
    properties: Dict[str, Any]
    required: List[str]


class McpTool(TypedDict, total=False):
    """MCP tool definition."""

    name: str
    description: str
    inputSchema: McpToolInputSchema


class McpClientConfig(TypedDict, total=False):
    """Configuration for MCP client."""

    api_key: str | None
    connection: str | None
    mode: str | None  # 'modelcontextprotocol' | 'toolkit'


class MatonMcpClient:
    """
    Client for connecting to the Maton MCP server at mcp.maton.ai.
    Fetches tool definitions and executes tool calls via MCP protocol.
    """

    def __init__(self, config: McpClientConfig):
        self._config = config
        self._tools: List[McpTool] = []
        self._initializer = AsyncInitializer()

        self._api_key = self._resolve_key(config.get("api_key"))

    def _resolve_key(self, key: str | None) -> str:
        """Resolve the API key, falling back to the MATON_API_KEY env var."""
        api_key = key or os.environ.get("MATON_API_KEY")
        if not api_key or not api_key.strip():
            raise ValueError(
                "Maton API key is required. Pass it as `api_key` or set the "
                "MATON_API_KEY environment variable. "
                "Create one at https://maton.ai."
            )
        return api_key

    def _get_headers(self) -> Dict[str, str]:
        """Build headers for MCP requests."""
        user_agent = (
            f"{MCP_HEADER}/{VERSION}"
            if self._config.get("mode") == "modelcontextprotocol"
            else f"{TOOLKIT_HEADER}/{VERSION}"
        )

        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "User-Agent": user_agent,
        }

        if self._config.get("connection"):
            headers["Maton-Connection"] = self._config["connection"]

        return headers

    @asynccontextmanager
    async def _create_session(self) -> AsyncGenerator[ClientSession, None]:
        """Create an MCP session within a proper async context.

        This ensures the connection lifecycle is managed correctly by
        using async with blocks, avoiding task group context issues.
        """
        headers = self._get_headers()

        async with streamablehttp_client(
            MCP_SERVER_URL,
            headers=headers,
            terminate_on_close=False
        ) as (read_stream, write_stream, _):
            async with ClientSession(read_stream, write_stream) as session:
                await session.initialize()
                yield session

    async def connect(self) -> None:
        """Connect to MCP server and fetch available tools."""
        await self._initializer.initialize(self._do_connect)

    async def _do_connect(self) -> None:
        """Internal connection logic."""
        try:
            async with self._create_session() as session:
                result = await session.list_tools()
                self._tools = [
                    McpTool(
                        name=t.name,
                        description=t.description or t.name,
                        inputSchema=t.inputSchema,
                    )
                    for t in result.tools
                ]
        except Exception as e:
            raise RuntimeError(
                f"Failed to connect to Maton MCP server at {MCP_SERVER_URL}. "
                f"Error: {str(e)}"
            ) from e

    @property
    def is_connected(self) -> bool:
        """Check if connected to MCP server."""
        return self._initializer.is_initialized

    def get_tools(self) -> List[McpTool]:
        """Get available tools. Must call connect() first."""
        if not self._initializer.is_initialized:
            raise RuntimeError(
                "MCP client not connected. "
                "Call connect() before accessing tools."
            )
        return self._tools

    async def call_tool(
        self,
        name: str,
        args: Dict[str, Any],
        connection: str | None = None
    ) -> str:
        """
        Execute a tool via MCP.

        Args:
            name: Tool method name (e.g., 'run_action')
            args: Tool arguments
            connection: Optional per-call connection override

        Returns:
            JSON string result
        """
        if not self._initializer.is_initialized:
            raise RuntimeError(
                "MCP client not connected. "
                "Call connect() before calling tools."
            )

        # Connection priority: per-call override > connection-time context > none
        final_connection = connection or self._config.get("connection")

        # Warn if args.connection exists and differs from override
        if (
            final_connection
            and args.get("connection")
            and args["connection"] != final_connection
        ):
            warnings.warn(
                f"[Maton Agent Toolkit] Connection context conflict detected:\n"
                f"  - Tool args.connection: {args['connection']}\n"
                f"  - Override connection: {final_connection}\n"
                f"  Using override connection. "
                f"This may indicate a bug in your code."
            )

        # Inject connection into args if present
        final_args = {**args}
        if final_connection:
            final_args["connection"] = final_connection

        try:
            async with self._create_session() as session:
                result = await session.call_tool(name, final_args)

                if result.isError:
                    error_text = next(
                        (
                            getattr(c, "text", None)
                            for c in result.content
                            if hasattr(c, "text")
                        ),
                        "Tool execution failed"
                    )
                    raise RuntimeError(str(error_text))

                # Extract text content
                text_content = next(
                    (
                        getattr(c, "text", None)
                        for c in result.content
                        if hasattr(c, "text")
                    ),
                    None
                )

                if text_content:
                    return text_content

                return json.dumps(result.model_dump())

        except Exception as e:
            raise RuntimeError(
                f"Failed to execute tool '{name}': {str(e)}"
            ) from e

    async def disconnect(self) -> None:
        """Disconnect from MCP server. Safe to call multiple times.

        Note: With this architecture, connections are opened and closed
        per-operation, so this just resets the initialized state.
        """
        if not self._initializer.is_initialized:
            return

        self._tools = []
        self._initializer.reset()

File: tools/typescript/src/shared/mcp-client.ts
import {Client} from '@modelcontextprotocol/sdk/client/index.js';
import {StreamableHTTPClientTransport} from '@modelcontextprotocol/sdk/client/streamableHttp.js';
import {AsyncInitializer} from './async-initializer';
import {VERSION, MCP_SERVER_URL, TOOLKIT_HEADER, MCP_HEADER} from './constants';

declare const process: {env: {MATON_API_KEY?: string}};

export interface McpClientConfig {
  apiKey?: string;
  context?: {
    connection?: string;
  };
  mode?: 'modelcontextprotocol' | 'toolkit';
}

export interface McpTool {
  name: string;
  description?: string;
  inputSchema?: {
    type: string;
    properties?: Record<string, unknown>;
    required?: string[];
  };
}

export interface McpToolCallResult {
  content: Array<{type: string; text?: string}>;
  isError?: boolean;
}

export class MatonMcpClient {
  private client: Client | null = null;
  private transport: StreamableHTTPClientTransport | null = null;
  private tools: McpTool[] = [];
  private apiKey: string;
  private config: McpClientConfig;
  private initializer = new AsyncInitializer();

  constructor(config: McpClientConfig) {
    this.config = config;
    this.apiKey = this.resolveKey(config.apiKey);
  }

  /**
   * Create transport and client fresh for each connection attempt.
   */
  private createTransportAndClient(): {
    transport: StreamableHTTPClientTransport;
    client: Client;
  } {
    // Determine User-Agent based on mode
    const userAgent =
      this.config.mode === 'modelcontextprotocol'
        ? `${MCP_HEADER}/${VERSION}`
        : `${TOOLKIT_HEADER}/${VERSION}`;

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
      'User-Agent': userAgent,
    };

    if (this.config.context?.connection) {
      headers['Maton-Connection'] = this.config.context.connection;
    }

    const transport = new StreamableHTTPClientTransport(
      new URL(MCP_SERVER_URL),
      {requestInit: {headers}}
    );

    const client = new Client(
      {
        name: TOOLKIT_HEADER,
        version: VERSION,
      },
      {
        capabilities: {},
      }
    );

    return {transport, client};
  }

  private resolveKey(key?: string): string {
    const apiKey = key || process.env.MATON_API_KEY;
    if (!apiKey || !apiKey.trim()) {
      throw new Error(
        'Maton API key is required. Pass it as `apiKey` or set the MATON_API_KEY environment variable. Create one at https://maton.ai.'
      );
    }
    return apiKey;
  }

  connect(): Promise<void> {
    return this.initializer.initialize(() => this.doConnect());
  }

  private async doConnect(): Promise<void> {
    try {
      // Create transport and client fresh for each connection attempt
      const {transport, client} = this.createTransportAndClient();
      this.transport = transport;
      this.client = client;

      await this.client.connect(this.transport);
      const result = await this.client.listTools();
      this.tools = result.tools as McpTool[];
    } catch (error) {
      // Clean up on failure
      this.client = null;
      this.transport = null;

      const errorMessage =
        error instanceof Error ? error.message : String(error);
      throw new Error(
        `Failed to connect to Maton MCP server at ${MCP_SERVER_URL}. ` +
          `Error: ${errorMessage}`,
        {cause: error}
      );
    }
  }

  isConnected(): boolean {
    return this.initializer.isInitialized;
  }

  getTools(): McpTool[] {
    if (!this.initializer.isInitialized) {
      throw new Error(
        'MCP client not connected. Call connect() before accessing tools.'
      );
    }
    return this.tools;
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
    options?: {connection?: string}
  ): Promise<string> {
    if (!this.initializer.isInitialized || !this.client) {
      throw new Error(
        'MCP client not connected. Call connect() before calling tools.'
      );
    }

    // Connection priority: per-call override > connection-time context > none.
    // When a connection is supplied, forward it as a `connection` argument so
    // the server routes the call through it (mirrors the Maton-Connection header).
    const connection = options?.connection ?? this.config.context?.connection;

    if (connection && args.connection && args.connection !== connection) {
      console.warn(
        `[Maton Agent Toolkit] Connection context conflict detected:\n` +
          `  - Tool args.connection: ${args.connection}\n` +
          `  - Override connection: ${connection}\n` +
          `  Using override connection. This may indicate a bug in your code.`
      );
    }

    const finalArgs = connection ? {...args, connection} : args;

    try {
      const result = (await this.client.callTool({
        name,
        arguments: finalArgs,
      })) as McpToolCallResult;

      if (result.isError) {
        const errorText = result.content?.find((c) => c.type === 'text')?.text;
        throw new Error(errorText || 'Tool execution failed');
      }

      const textContent = result.content?.find((c) => c.type === 'text');
      if (textContent && textContent.text) {
        return textContent.text;
      }

      return JSON.stringify(result);
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      throw new Error(`Failed to execute tool '${name}': ${errorMessage}`, {
        cause: error,
      });
    }
  }

  /**
   * Disconnect from the MCP server and clean up resources.
   * Safe to call multiple times (idempotent).
   */
  async disconnect(): Promise<void> {
    if (!this.initializer.isInitialized) {
      return; // Already disconnected or never connected
    }

    try {
      if (this.client) {
        await this.client.close();
      }
    } finally {
      // Always clean up state, even if close() throws
      this.client = null;
      this.transport = null;
      this.tools = [];
      this.initializer.reset();
    }
  }
}

export default MatonMcpClient;
